core/src/test/java/org/elasticsearch/action/admin/indices/mapping/put/PutMappingRequestTests.java

```diff
@@ -64,4 +71,42 @@ public void testBuildFromSimplifiedDef() {
                 () -> PutMappingRequest.buildFromSimplifiedDef("type", "only_field"));
         assertEquals("mapping source must be pairs of fieldnames and properties definition.", e.getMessage());
     }
+
+    public void testPutMappingRequestFromOldVersion() throws IOException {
+        // this is hacky but here goes
+        PutMappingRequest request = new PutMappingRequest("foo");
+        String mapping = YamlXContent.contentBuilder().startObject().field("foo", "bar").endObject().string();
+        request.source(mapping, XContentType.JSON); // THIS IS NOT A BUG! Intentionally specifying the wrong type so we serialize it
+        assertEquals(mapping, request.source());
+
+        // output version doesn't matter
+        BytesStreamOutput bytesStreamOutput = new BytesStreamOutput();
+        request.writeTo(bytesStreamOutput);
+
+        StreamInput in = StreamInput.wrap(bytesStreamOutput.bytes().toBytesRef().bytes);
+        in.setVersion(Version.V_5_0_0);
```

Comment:
in all these serialization tests, shall we randomize the version just to make sure that the conditional target the proper versions? you may bump into the fact that some released versions are still marked unreleased hence cannot be used in VersionUtils.randomVersion.